### PR TITLE
[ShapesSpec] Plumb ShapesSpec / ParamsSpec through strict torch.export + unbacked for export finally. 

### PR DESCRIPTION
**Pull request by laithsakka**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #187010
* #186751
* #187160
* #186167
* __->__ #185982

How does it work?

Export's signature order of arguments depends on the args and kwargs passed to the export call: export creates an intermediate module that takes variadic args as input. ShapesSpec, on the other hand, is defined against the exported function's signature. To reconcile the two, we also build an intermediate shape spec (this is what the flatten-spec function does).

Derived expressions, assumptions, and hints have been tested and work fine.

We do not support containers yet (objects, dicts, lists) — that will come in a follow-up PR.
```python
from torch.fx.experimental.dynamic_spec import (
    ParamsSpec as PARAMS,
    ShapesSpec,
    ShapeVar as VAR,
    TensorSpec as T,
)

batch = VAR("batch", min=2, max=128)
ep = torch.export.export(
    mod,
    (torch.randn(8, 3), torch.randn(16, 3)),
    dynamic_shapes=ShapesSpec(
        params=PARAMS(
            {
                "x": T([batch, 3]),
                "y": T([batch * 2, 3]),  # derived expression
            }
        ),
        assumptions=[batch % 2 == 0],
    ),
    strict=True,
)
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

File: torch/export/__init__.py [hunk 1]
```diff
@@ -127,6 +128,44 @@ def export(
...
+         full details):
+
+         * **Unbacked-only.** Dims / scalars marked dynamic become unbacked
+           SymInts (``u`` symbols) and are never specialized (including no
+           0/1 specialization).
+         * **Assumptions and derived expressions.** You may attach
+           assumptions and expressions derived from the spec's symbols.
+         * **Export-time soundness.** The exported graph is guaranteed valid
+           for every assumption provided, otherwise export fails. (The legacy
+           ``Dim.DYNAMIC`` / ``Dim.AUTO`` mode may silently specialize a
```
Comment (angelayi) [inline torch/export/__init__.py]: There is a user base for Dim.DYNAMIC/AUTO (those who don't want to figure out dynamic constraints), so I would want to keep supporting this, instead of calling it a "legacy" solution, unless ShapeSpec can support it

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by jansel** [comment]
@claude review these changes

**Comment by laithsakka** [comment]
> **Claude finished @jansel's task in 2m 33s** —— [View job](https://github.com/pytorch/pytorch/actions/runs/26969051140)
> 
> ### Code Review: ShapesSpec / ParamsSpec plumbing through strict export
...

1. lgtm addressed
2. I want it like this on purpose it will break internal users otherwise due to added public dependencies. 
3. Adding a helper function is longer than inlining the check i will leave it as is i will just store the error text in a global variable to avoid duplicating it. 
4. i think its fine as is
5. lgtm addressed. 

nits:
i think nothing worth addressing.

File: torch/export/_trace.py [hunk 2]
```diff
@@ -115,8 +120,16 @@
 
 log = logging.getLogger(__name__)
 
-# Type alias for dynamic shapes specification
-_DynamicShapesSpec: TypeAlias = dict[str, Any] | tuple[Any, ...] | list[Any]
+# Legacy dynamic shapes specification: the raw container spelling that
+# predates the structured ShapesSpec / ParamsSpec types. AdditionalInputs
+# and ShapesCollection are entry-point helpers that get unwrapped into one
```
Comment (angelayi) [inline torch/export/_trace.py]: do you plan to support AdditionalInputs/ShapesCollection -> ShapeSpec?

File: torch/export/_trace.py [hunk 3]
```diff
@@ -588,6 +601,44 @@ def _replace_unbacked_bindings(gm: torch.fx.GraphModule) -> None:
...
+    from torch.fx.experimental.symbolic_shapes import (
+        _free_unbacked_symbols_with_path,
+        free_unbacked_symbols,
+    )
+
+    if (shape_env := _get_shape_env_from_gm(gm)) is None:
+        return
+
+    for node in gm.graph.nodes:
+        if node.op != "placeholder" or node.meta.get("unbacked_bindings"):
```
Comment (angelayi) [inline torch/export/_trace.py]:
```suggestion
    for node in gm.graph.find_nodes(op="placeholder"):
        if node.meta.get("unbacked_bindings"):
```

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

File: test/export/test_dynamic_spec_export.py [hunk 4]
```diff
@@ -0,0 +1,896 @@
...
+        )
+
+        # Sanity-check: the exported program runs with different shapes.
+        ep.module()(torch.randn(16, 3))
+        ep.module()(torch.randn(32, 3))
+
+    def test_static_int_spec_mismatch_raises(self):
+        with self.assertRaisesRegex(
+            ValueError,
+            r"shapes_spec declares L\['flat_args'\]\[1\] as static with value 10, but got 42",
```
Comment (angelayi) [inline test/export/test_dynamic_spec_export.py]: nit: 
```suggestion
            r"shapes_spec declared L\['flat_args'\]\[1\] as static with value 10, but while tracing we found that it was actually 42",
```

File: test/export/test_dynamic_spec_export.py [hunk 5]
```diff
@@ -0,0 +1,896 @@
...
+            def forward(self, x, n):
+                return x.sum(0) * n
+
+        ep = export(
+            M(),
+            (torch.randn(8, 3), 5),
+            dynamic_shapes=ParamsSpec(
+                {
+                    "x": TensorSpec([ShapeVar("B"), STATIC]),
+                    "n": IntVar("n_size"),
```
Comment (angelayi) [inline test/export/test_dynamic_spec_export.py:270]: does that make this integer dynamic? can we have a test to show this being dynamic

File: test/export/test_dynamic_spec_export.py [hunk 6]
```diff
@@ -0,0 +1,886 @@
...
+
+
+class TestExportDynamicSpec(TestCase):
+    """torch.export.export support for the new ShapesSpec/ParamsSpec API."""
+
+    def setUp(self):
+        super().setUp()
+        _reset_uid_counter()
+
+    def test_unbacked_graph_has_unbacked_symbol(self):
```
Comment (angelayi) [inline test/export/test_dynamic_spec_export.py:92]: can we somehow test through all the existing named-Dims test cases in test_export? by converting the named Dims to the ShapeSpec or something?

**Review by jansel — APPROVED** [review]

File: torch/export/__init__.py (hunk 1 above)
Comment (laithsakka) [reply torch/export/__init__.py]: I will remove the legacy terms, but they are NOT sounds approaches though --> meaning when used users are ok to export even if there are some silent constraints in the graph.

that said n general i do want people to move away from `Dim.DYNAMIC`` / ``Dim.AUTO`` tbh

File: test/export/test_dynamic_spec_export.py (hunk 5 above)
Comment (laithsakka) [reply test/export/test_dynamic_spec_export.py:270]: yeh the test bellow does check. 
is that not enough?
```
        self.assertRegex(ep_str, r'n: "Sym\(u\d+\)"')
```

Comment (laithsakka) [reply test/export/test_dynamic_spec_export.py:270]: would you rather i print whole graph here?

Comment (laithsakka) [reply test/export/test_dynamic_spec_export.py:270]: you can do  "n": IntVar("n_size"),or  "n": VAR("n_size") btw VAR -> SHAPE VAR (0->inf)
intVar [-inf->inf]

File: test/export/test_dynamic_spec_export.py (hunk 6 above)
Comment (laithsakka) [reply test/export/test_dynamic_spec_export.py:92]: mmm not bad idea, graphs will be different u0, vs s0 .. 
lmk look into this and see if its worth it or not.

File: torch/export/__init__.py (hunk 1 above)
Comment (laithsakka) [reply torch/export/__init__.py]: ok addressed by updating comment and not calling it legacy

File: torch/export/_trace.py (hunk 2 above)
Comment (laithsakka) [reply torch/export/_trace.py]: mm when the need arise yes, but not now. i will add a coment that  AdditionalInputs/ShapesCollection only works with the Dim API

File: torch/export/_trace.py (hunk 3 above)
Comment (laithsakka) [reply torch/export/_trace.py]: done

File: test/export/test_dynamic_spec_export.py (hunk 4 above)
Comment (laithsakka) [reply test/export/test_dynamic_spec_export.py]: done

File: test/export/test_dynamic_spec_export.py (hunk 6 above)
Comment (laithsakka) [reply test/export/test_dynamic_spec_export.py:92]: if we are going todo this it actually better be done in separate PR.

File: test/export/test_dynamic_spec_export.py (hunk 5 above)
Comment (angelayi) [reply test/export/test_dynamic_spec_export.py:270]: oh wait sorry i misread the test cases below it looks good :D

File: torch/export/__init__.py (hunk 1 above)
Comment (angelayi) [reply torch/export/__init__.py]: Dim.DYNAMIC/AUTO should be sound in that if users pass in something that violates an implicit constraint created at tracing time, the runtime should properly error

File: torch/export/_trace.py (hunk 2 above)
Comment (angelayi) [reply torch/export/_trace.py]: ShapesCollection is very much used when marking KJTs dynamic

**Review by angelayi — APPROVED** [review]
